File: app/src/main/java/com/vondear/tools/view/RxDialogShopCart.java
package com.vondear.tools.view;

import android.animation.Animator;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.vondear.tools.R;
import com.vondear.tools.adapter.AdapterPopupDish;
import com.vondear.tools.model.ModelShopCart;
import com.vondear.tools.interfaces.ShopCartInterface;


/**
 * @author vondear
 * @date 16-12-22
 */
public class RxDialogShopCart extends Dialog implements View.OnClickListener, ShopCartInterface {

    private LinearLayout linearLayout, bottomLayout, clearLayout;
    private FrameLayout shopingcartLayout;
    private ModelShopCart mModelShopCart;
    private TextView totalPriceTextView;
    private TextView totalPriceNumTextView;
    private RecyclerView recyclerView;
    private AdapterPopupDish dishAdapter;
    private ShopCartDialogImp shopCartDialogImp;

    public RxDialogShopCart(Context context, ModelShopCart modelShopCart, int themeResId) {
        super(context, themeResId);
        this.mModelShopCart = modelShopCart;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.cart_popupview);
        linearLayout = findViewById(R.id.linearlayout);
        clearLayout = findViewById(R.id.clear_layout);
        shopingcartLayout = findViewById(R.id.shopping_cart_layout);
        bottomLayout = findViewById(R.id.shopping_cart_bottom);
        totalPriceTextView = findViewById(R.id.shopping_cart_total_tv);
        totalPriceNumTextView = findViewById(R.id.shopping_cart_total_num);
        recyclerView = findViewById(R.id.recycleview);
        shopingcartLayout.setOnClickListener(this);
        bottomLayout.setOnClickListener(this);
        clearLayout.setOnClickListener(this);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        dishAdapter = new AdapterPopupDish(getContext(), mModelShopCart);
        recyclerView.setAdapter(dishAdapter);
        dishAdapter.setShopCartInterface(this);
        showTotalPrice();
    }

    @Override
    public void show() {
        super.show();
        animationShow(500);
    }

    @Override
    public void dismiss() {
        animationHide(500);
    }

    private void showTotalPrice() {
        if (mModelShopCart != null && mModelShopCart.getShoppingTotalPrice() > 0) {
            totalPriceTextView.setVisibility(View.VISIBLE);
            totalPriceTextView.setText(getContext().getResources().getString(R.string.rmb) + " " + mModelShopCart.getShoppingTotalPrice());
            totalPriceNumTextView.setVisibility(View.VISIBLE);
            totalPriceNumTextView.setText("" + mModelShopCart.getShoppingAccount());

        } else {
            totalPriceTextView.setVisibility(View.GONE);
            totalPriceNumTextView.setVisibility(View.GONE);
        }
    }

    private void animationShow(int mDuration) {
        AnimatorSet animatorSet = new AnimatorSet();
        animatorSet.playTogether(
                ObjectAnimator.ofFloat(linearLayout, "translationY", 1000, 0).setDuration(mDuration)
        );
        animatorSet.start();
    }

    private void animationHide(int mDuration) {
        AnimatorSet animatorSet = new AnimatorSet();
        animatorSet.playTogether(
                ObjectAnimator.ofFloat(linearLayout, "translationY", 0, 1000).setDuration(mDuration)
        );
        animatorSet.start();

        if (shopCartDialogImp != null) {
            shopCartDialogImp.dialogDismiss();
        }

        animatorSet.addListener(new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animator) {
            }

            @Override
            public void onAnimationEnd(Animator animator) {
                RxDialogShopCart.super.dismiss();
            }

            @Override
            public void onAnimationCancel(Animator animator) {

            }

            @Override
            public void onAnimationRepeat(Animator animator) {

            }
        });
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.shopping_cart_bottom:
            case R.id.shopping_cart_layout:
                this.dismiss();
                break;
            case R.id.clear_layout:
                clear();
                break;
        }
    }

    @Override
    public void add(View view, int postion) {
        showTotalPrice();
    }

    @Override
    public void remove(View view, int postion) {
        showTotalPrice();
        if (mModelShopCart.getShoppingAccount() == 0) {
            this.dismiss();
        }
    }

    public ShopCartDialogImp getShopCartDialogImp() {
        return shopCartDialogImp;
    }

    public void setShopCartDialogImp(ShopCartDialogImp shopCartDialogImp) {
        this.shopCartDialogImp = shopCartDialogImp;
    }

    public interface ShopCartDialogImp {
        public void dialogDismiss();
    }

    public void clear() {
        mModelShopCart.clear();
        showTotalPrice();
        if (mModelShopCart.getShoppingAccount() == 0) {
            this.dismiss();
        }
    }
}

File: RxDemo/src/main/java/com/tamsiree/rxdemo/view/RxDialogShopCart.kt
package com.tamsiree.rxdemo.view

import android.animation.Animator
import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.annotation.SuppressLint
import android.app.Dialog
import android.content.Context
import android.os.Bundle
import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import com.tamsiree.rxdemo.R
import com.tamsiree.rxdemo.adapter.AdapterPopupDish
import com.tamsiree.rxdemo.interfaces.ShopCartInterface
import com.tamsiree.rxdemo.model.ModelShopCart
import kotlinx.android.synthetic.main.cart_popupview.*

/**
 * @author tamsiree
 * @date 16-12-22
 */
class RxDialogShopCart(context: Context?, private val mModelShopCart: ModelShopCart?, themeResId: Int) : Dialog(context!!, themeResId), View.OnClickListener, ShopCartInterface {

    var shopCartDialogImp: ShopCartDialogImp? = null
    override fun onCreate(savedInstanceState: Bundle) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.cart_popupview)

        shopping_cart_layout?.setOnClickListener(this)
        shopping_cart_bottom?.setOnClickListener(this)
        clear_layout.setOnClickListener(this)
        recycleview.layoutManager = LinearLayoutManager(context)
        val dishAdapter = AdapterPopupDish(context, mModelShopCart!!)
        recycleview.adapter = dishAdapter
        dishAdapter.shopCartInterface = this
        showTotalPrice()
    }

    override fun show() {
        super.show()
        animationShow(500)
    }

    override fun dismiss() {
        animationHide(500)
    }

    @SuppressLint("SetTextI18n")
    private fun showTotalPrice() {
        if (mModelShopCart != null && mModelShopCart.shoppingTotalPrice > 0) {
            shopping_cart_total_tv!!.visibility = View.VISIBLE
            shopping_cart_total_tv!!.text = context.resources.getString(R.string.rmb) + " " + mModelShopCart.shoppingTotalPrice
            shopping_cart_total_num!!.visibility = View.VISIBLE
            shopping_cart_total_num!!.text = "" + mModelShopCart.shoppingAccount
        } else {
            shopping_cart_total_tv!!.visibility = View.GONE
            shopping_cart_total_num!!.visibility = View.GONE
        }
    }

    private fun animationShow(mDuration: Int) {
        val animatorSet = AnimatorSet()
        animatorSet.playTogether(
                ObjectAnimator.ofFloat(linearlayout, "translationY", 1000f, 0f).setDuration(mDuration.toLong())
        )
        animatorSet.start()
    }

    private fun animationHide(mDuration: Int) {
        val animatorSet = AnimatorSet()
        animatorSet.playTogether(
                ObjectAnimator.ofFloat(linearlayout, "translationY", 0f, 1000f).setDuration(mDuration.toLong())
        )
        animatorSet.start()
        if (shopCartDialogImp != null) {
            shopCartDialogImp!!.dialogDismiss()
        }
        animatorSet.addListener(object : Animator.AnimatorListener {
            override fun onAnimationStart(animator: Animator) {}
            override fun onAnimationEnd(animator: Animator) {
                super@RxDialogShopCart.dismiss()
            }

            override fun onAnimationCancel(animator: Animator) {}
            override fun onAnimationRepeat(animator: Animator) {}
        })
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.shopping_cart_bottom, R.id.shopping_cart_layout -> dismiss()
            R.id.clear_layout -> clear()
        }
    }

    override fun add(view: View?, position: Int) {
        showTotalPrice()
    }

    override fun remove(view: View?, position: Int) {
        showTotalPrice()
        if (mModelShopCart!!.shoppingAccount == 0) {
            dismiss()
        }
    }

    interface ShopCartDialogImp {
        fun dialogDismiss()
    }

    fun clear() {
        mModelShopCart!!.clear()
        showTotalPrice()
        if (mModelShopCart.shoppingAccount == 0) {
            dismiss()
        }
    }

}